Add monitoring support for hardware accelerators

rohitagarwal003 opened the pull request:
Currently only NVIDIA GPU monitoring is implemented.

Feature repo issue: https://github.com/kubernetes/features/issues/369
cAdvisor PR: https://github.com/google/cadvisor/pull/1762

/kind feature
/sig node
/sig instrumentation
/area hw-accelerators

**Release note**:
```release-note
Kubelet now exposes metrics for NVIDIA GPUs attached to the containers.
```

rohitagarwal003 committed:
Update cAdvisor.

Also update golang.org/x/sys because of google/cadvisor#1786

rohitagarwal003 committed:
Expose accelerator metrics in the summary API.

mindprince commented:
/assign @vishh @yujuhong

vishh commented on pkg/kubelet/apis/stats/v1alpha1/types.go:
```diff
@@ -188,6 +190,29 @@ type MemoryStats struct {
+	// unit: bytes
+	MemoryUsed uint64 `json:"memory_used"`
+
+	// Percent of time over the past sample period during which
```
Define what the sampling period is ? It is `10 seconds` right?

mindprince replied:
Done.

vishh commented:
This PR LGTM. I'd recommend making the cAdvisor update in a separate PR though.
This PR can have just the summary API changes, ideally with corresponding e2es updates as well.

cmingxu commented:
/retest

rohitagarwal003 committed:
Fix TestSummaryProvider.

assert.Contains() checks if its second argument (which is supposed to be
a single element) is contained in its first argument (which is supposed
to be a slice/map etc.) The third and following arguments are supposed
to be message and args for the output in case of failure.

Because of this bad form, a failure was hidden, the system container is
named "misc", not "system".

andyxning commented:
@mindprince @vishh I can help to add GPU info to heapster based on this.

BTW, nice work~

mindprince commented:
/assign @dchen1107 
This is ready to go in.

abdasgupta commented:
@mindprince  One thing for  my clarification. Will this PR show the GPU metrics on the following endpoint: http://<KUBELET-IP>:10255/stats/summary ?

mindprince commented:
@abdasgupta Yes.

mindprince commented:
/cc @dashpole @Random-Liu @yguo0905 
Adding more people from sig node who have worked with the summary API.

yguo0905 commented on pkg/kubelet/server/stats/summary_test.go:
```diff
@@ -46,7 +46,7 @@ func TestSummaryProvider(t *testing.T) {
 		nodeConfig   = cm.NodeConfig{
 			RuntimeCgroupsName: "/runtime",
-			SystemCgroupsName:  "/system",
+			SystemCgroupsName:  "/misc",
```
Just curious why do we need this change?

mindprince replied:
I explain this here: https://github.com/kubernetes/kubernetes/pull/55188/commits/238b4a0d8e059e6b648fbb94d324b183fe1dc614

```
Fix TestSummaryProvider.

assert.Contains() checks if its second argument (which is supposed to be
a single element) is contained in its first argument (which is supposed
to be a slice/map etc.) The third and following arguments are supposed
to be message and args for the output in case of failure.

Because of this bad form, a failure was hidden, the system container is
named "misc", not "system".
```

yguo0905 reviewed (commented):
LGTM

dchen1107 commented:
I only reviewed the last commit: Expose accelerator metrics in the summary API.

/lgtm

dchen1107 commented:
/approve

mindprince commented:
/retest

andyxning commented:
Any thing wrong. Seems submit queue is pending.

smarterclayton commented:
How was this merged without a proposal merged to kubernetes/community?

smarterclayton commented on pkg/kubelet/apis/stats/v1alpha1/types.go:
```diff
@@ -188,6 +190,30 @@ type MemoryStats struct {
+
+	// Percent of time over the past sample period (10s) during which
+	// the accelerator was actively processing.
+	DutyCycle uint64 `json:"duty_cycle"`
```
This is not consistent with our API conventions - use of snake_case is not allowed in public APIs.

smarterclayton commented on pkg/kubelet/apis/stats/v1alpha1/types.go:
```diff
@@ -188,6 +190,30 @@ type MemoryStats struct {
+
+	// Total accelerator memory.
+	// unit: bytes
+	MemoryTotal uint64 `json:"memory_total"`
```
This is not consistent with our API conventions - use of snake_case is not allowed in public APIs.

smarterclayton commented on pkg/kubelet/apis/stats/v1alpha1/types.go:
```diff
@@ -188,6 +190,30 @@ type MemoryStats struct {
+
+	// Total accelerator memory allocated.
+	// unit: bytes
+	MemoryUsed uint64 `json:"memory_used"`
```
This is not consistent with our API conventions - use of snake_case is not allowed in public APIs.

smarterclayton commented:
Looks like there are some API errors on this - @mindprince please open a new PR for 1.9 that fixes the API to be covered by our API conventions.

smarterclayton commented:
@kubernetes/sig-testing-feature-requests I think I need to make sure `pkg/kubelet/apis` appears under the API changes label (it wasn't flagged here and so I missed it).  Can someone point me to where that can be done?

mindprince commented:
@smarterclayton https://github.com/kubernetes/kubernetes/pull/55908

vishh commented:
@smarterclayton can we build some CI tooling to enforce these semantics?
 I feel a full design review just for the sake of API schema review is heavy weight. 
The design doc was shared as a google doc and can now be committed to k8s/community repo if that will be useful.
However, it is not clear if the entire golang object definition is expected in design docs though.
